stan/math/rev/mat/fun/matrix_power.hpp

```diff
@@ -0,0 +1,182 @@
+    Mnd = Eigen::Matrix<double, R, C>::Identity(M.rows(), M.cols());
+    Eigen::Map<matrix_vi>(adjMnRef_, M.rows(), M.cols())
+        = Mnd.unaryExpr([](double x) { return new vari(x, false); });
+  }
+
+  virtual void chain() { return; }
+};
+
+template <int R, int C>
+class matrix_product_vari_n1 : public vari {
+ public:
+  int rows_;
+  int cols_;
+  double* M_;
+  vari** adjMRef_;
+  vari** adjMnRef_;
+
+  explicit matrix_product_vari_n1(const Eigen::Matrix<var, R, C>& M)
+      : vari(0.0),
+        rows_(M.rows()),
+        cols_(M.cols()),
+        M_(reinterpret_cast<double*>(ChainableStack::instance_->memalloc_.alloc(
+            sizeof(double) * M.rows() * M.cols()))),
+        adjMRef_(
+            reinterpret_cast<vari**>(ChainableStack::instance_->memalloc_.alloc(
+                sizeof(vari*) * M.rows() * M.cols()))),
+        adjMnRef_(
+            reinterpret_cast<vari**>(ChainableStack::instance_->memalloc_.alloc(
+                sizeof(vari*) * M.rows() * M.cols()))) {
+    Eigen::Map<matrix_vi>(adjMRef_, rows_, cols_) = M.vi();
+    Eigen::Map<matrix_d> Md(M_, M.rows(), M.cols());
+    Md = M.val();
+    Eigen::Map<matrix_vi>(adjMnRef_, M.rows(), M.cols())
+        = Md.unaryExpr([](double x) { return new vari(x, false); });
+  }
+
+  virtual void chain() {
+    Eigen::Map<matrix_vi>(adjMRef_, rows_, cols_).adj()
+        = Eigen::Map<matrix_vi>(adjMnRef_, rows_, cols_).adj();
+  }
+};
+
+template <int R, int C>
+class matrix_product_vari : public vari {
+ public:
+  int rows_;
+  int cols_;
+  vari** adjMRef_;
+  vari** adjMnRef_;
+  std::vector<Eigen::Matrix<double, R, C> > Mds;  // M, M^2, ..., M^(n-1), M^n.
+  int n;
+
+  explicit matrix_product_vari(const Eigen::Matrix<var, R, C>& M, const int n)
+      : vari(0.0),
+        rows_(M.rows()),
+        cols_(M.cols()),
+        adjMRef_(
+            reinterpret_cast<vari**>(ChainableStack::instance_->memalloc_.alloc(
+                sizeof(vari*) * M.rows() * M.cols()))),
+        adjMnRef_(
+            reinterpret_cast<vari**>(ChainableStack::instance_->memalloc_.alloc(
+                sizeof(vari*) * M.rows() * M.cols()))),
+        n(n) {
+    Mds.resize(n);
+    Mds[0] = M.val();
+    Eigen::Map<matrix_vi>(adjMRef_, rows_, cols_) = M.vi();
+    for (int i = 1; i < n; i++) {
+      Mds[i] = Mds[i - 1] * Mds[0];
+    }
+    Eigen::Map<matrix_vi>(adjMnRef_, rows_, cols_)
+        = Mds[n - 1].unaryExpr([](double x) { return new vari(x, false); });
+  }
+
+  virtual void chain() {
+    for (int i = 0; i < rows_; i++) {
+      for (int j = 0; j < cols_; j++) {
+        double dijdkl = 0.0;
+        Eigen::Matrix<double, R, C> S
+            = Eigen::Matrix<double, R, C>::Zero(rows_, cols_);
+        for (int l = 0; l < cols_; l++) {
+          S(i, l) += Mds[n - 2](j, l);
+        }
+        for (int nn = 1; nn < n - 1; nn++) {
+          for (int k = 0; k < rows_; k++) {
+            for (int l = 0; l < cols_; l++) {
```

Comment:
Matrices are column major, so the loops should go the other way to preserve memory locality.

Reply:
Loops swapped.